substrate-node/pallets/pallet-smart-contract/src/lib.rs

```diff
@@ -534,68 +675,163 @@ use sp_std::convert::{TryFrom, TryInto};
+
+        Ok(().into())
+    }
+
+    #[transactional]
+    pub fn _create_capacity_reservation_contract(
+        account_id: T::AccountId,
+        farm_id: u32,
+        policy: CapacityReservationPolicy,
         solution_provider_id: Option<u64>,
     ) -> DispatchResultWithPostInfo {
         let twin_id = pallet_tfgrid::TwinIdByAccountID::<T>::get(&account_id)
             .ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            pallet_tfgrid::Farms::<T>::contains_key(farm_id),
+            Error::<T>::FarmNotExists
+        );
 
-        let node = pallet_tfgrid::Nodes::<T>::get(node_id).ok_or(Error::<T>::NodeNotExists)?;
-        let farm = pallet_tfgrid::Farms::<T>::get(node.farm_id).ok_or(Error::<T>::FarmNotExists)?;
-
-        if farm.dedicated_farm && !ActiveRentContractForNode::<T>::contains_key(node_id) {
-            return Err(Error::<T>::NodeNotAvailableToDeploy.into());
-        }
-
-        // If the user is trying to deploy on a node that has an active rent contract
-        // only allow the user who created the rent contract to actually deploy a node contract on it
-        if let Some(contract_id) = ActiveRentContractForNode::<T>::get(node_id) {
-            let rent_contract =
-                Contracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExists)?;
-            if rent_contract.twin_id != twin_id {
-                return Err(Error::<T>::NodeHasRentContract.into());
+        let (node_id, resources, group_id) = match policy {
+            CapacityReservationPolicy::Any {
+                resources,
+                features,
+            } => {
+                let farm =
+                    pallet_tfgrid::Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+                ensure!(!farm.dedicated_farm, Error::<T>::NodeNotAvailableToDeploy);
+                let n_id = Self::_find_suitable_node_in_farm(farm_id, resources, None, features)?;
+                (n_id, resources, None)
             }
-        }
-
-        // If the contract with hash and node id exists and it's in any other state then
-        // contractState::Deleted then we don't allow the creation of it.
-        // If it exists we allow the user to "restore" this contract
-        if ContractIDByNodeIDAndHash::<T>::contains_key(node_id, &deployment_hash) {
-            let contract_id = ContractIDByNodeIDAndHash::<T>::get(node_id, &deployment_hash);
-            let contract = Contracts::<T>::get(contract_id).ok_or(Error::<T>::ContractNotExists)?;
-            if !contract.is_state_delete() {
-                return Err(Error::<T>::ContractIsNotUnique.into());
+            CapacityReservationPolicy::Exclusive {
+                group_id,
+                resources,
+                features,
+            } => {
+                let farm =
+                    pallet_tfgrid::Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+                ensure!(!farm.dedicated_farm, Error::<T>::NodeNotAvailableToDeploy);
```

Comment:
why is this check here for creating a capacity reservation contract on a dedicated farm? It should still be possible to do this.

Reply:
Do I add the extra resources check (checking if the passed resources equals the total resources of the node) here?